// expressio/lib/mailer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /**
                                                                                                                                                                                                                                                                   * Mailer
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * @copyright Copyright (c) 2017, hugw.io
                                                                                                                                                                                                                                                                   * @author Hugo W - contact@hugw.io
                                                                                                                                                                                                                                                                   * @license MIT
                                                                                                                                                                                                                                                                   */

var _nodemailer = require('nodemailer');

var _nodemailer2 = _interopRequireDefault(_nodemailer);

var _joi = require('joi');

var _joi2 = _interopRequireDefault(_joi);

var _utils = require('./utils');

var _utils2 = _interopRequireDefault(_utils);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Object schemas
 * to validate configuration
 */
const schema = _joi2.default.object({
  enabled: _joi2.default.boolean().required(),
  transport: _joi2.default.object().required(),
  defaults: _joi2.default.object({
    from: _joi2.default.string().required()
  }).required()
});

exports.default = (server, config) => {
  const {
    enabled,
    transport,
    defaults
  } = _utils2.default.sanitize(config, schema, 'Invalid Mailer config');

  if (!enabled) return;
  const transporter = _nodemailer2.default.createTransport(transport, defaults);
  const { logger } = server;

  /**
   * Dispatch email
   */
  transporter.dispatch = async msg => {
    try {
      const info = await transporter.sendMail(msg);

      const response = _extends({}, info, {
        content: msg.text || msg.html || ''
      });

      const debug = _nodemailer2.default.getTestMessageUrl(info) || null;

      logger.info(null, {
        mailer: { response, debug },
        options: logger.options
      });

      return info;
    } catch (err) {
      err.message = `Mailer error: ${err.message}`;
      throw err;
    }
  };

  // Expose Mailer to the server object
  server.mailer = transporter;

  // Expose Mailer to the request object
  server.use((req, res, next) => {
    req.mailer = transporter;
    next();
  });
};
